consensus/istanbul/backend/announce.go

```diff
@@ -370,52 +369,42 @@ func (sb *Backend) generateAndGossipAnnounce(version uint) error {
 		return err
 	}
 
-	// Add the generated announce message to this node's cache
-	sb.cachedAnnounceMsgsMu.Lock()
-	defer sb.cachedAnnounceMsgsMu.Unlock()
-	sb.cachedAnnounceMsgs[announceAddress] = &announceMsgCachedEntry{MsgVersion: version, MsgPayload: payload}
-
 	return sb.Multicast(nil, payload, istanbulAnnounceMsg)
 }
 
-// This function is a helper function for generateAndGossipAnnounce.
-// It will create the latest announce msg for this node with a given version.
-func (sb *Backend) generateAnnounce(version uint) (*istanbul.Message, common.Address, error) {
+// This function is a helper function for generateAndGossipAnnounce that will
+// use the version for this node's address in the signed announce version db
+func (sb *Backend) generateAnnounce() (*istanbul.Message, error) {
 	logger := sb.logger.New("func", "generateAnnounce")
-	var enodeUrl string
-	if sb.config.Proxied {
-		if sb.proxyNode != nil {
-			enodeUrl = sb.proxyNode.externalNode.URLv4()
-		} else {
-			logger.Error("Proxied node is not connected to a proxy")
-			return nil, common.Address{}, errNoProxyConnection
-		}
-	} else {
-		enodeUrl = sb.p2pserver.Self().URLv4()
-	}
 
-	// Retrieve the set of remote validators' public key to encrypt the enodeUrl
-	validatorConnSet, err := sb.retrieveValidatorConnSet()
+	enodeURL, err := sb.getEnodeURL()
 	if err != nil {
-		return nil, common.Address{}, err
+		logger.Error("Error getting enode URL", "err", err)
+		return nil, err
 	}
-
-	announceRecords := make([]*announceRecord, 0, len(validatorConnSet))
-	for addr := range validatorConnSet {
-		// TODO - Need to encrypt using the remote validator's validator key
-		announceRecords = append(announceRecords, &announceRecord{DestAddress: addr, EncryptedEnodeURL: []byte(enodeUrl)})
+	announceRecords, err := sb.generateAnnounceRecords(enodeURL)
+	if err != nil {
+		logger.Warn("Error generating announce records", "err", err)
+		return nil, err
+	}
+	if len(announceRecords) == 0 {
+		logger.Trace("No announce records were generated, will not generate announce")
+		return nil, nil
+	}
+	announceVersion, err := sb.signedAnnounceVersionTable.GetVersion(sb.Address())
```

Comment:
Is it possible that we could get stuck here if the entry does not exist or the DB gets corrupted? This may need a fallback for if it returns an error.

Reply:
Good point, changed this to rely on `announceVersion` in the `announceThread`